### ISSUE TITLE
如何在bert4torch下使用tensorboard ？

### ISSUE THREAD

**Issue by XuJianzhi** [Tongjilibo/bert4torch#58]
波哥好。

我知道torch能加tensorboard，
from torch.utils.tensorboard import SummaryWriter
writer = SummaryWriter(path)
.......
writer.add_scalar('train_loss', train_loss, epoch)

bert4torch在训练过程中能打印出，但如何以变量形式拿到loss和val_f1等信息？

如果暂时没有的话，，，我想自己改动，，，类似这样，应该改哪里的代码？
loss_list, prec_list, recall_list, f1_list = model.fit(train_dataloader, ... )

感谢大佬

**Comment by XuJianzhi** [comment]
是不是只能从这里拿到？  https://github.com/Tongjilibo/torch4keras/blob/master/torch4keras/model.py#L200

**Comment by Tongjilibo** [comment]
有默认的tensorboard回调，目前实现逻辑抽象到torch4keras中了，bert4torch使用方式和之前一样，参考https://github.com/Tongjilibo/bert4torch/blob/master/examples/tutorials/tutorials_small_tips.py

**Comment by XuJianzhi** [comment]
> 有默认的tensorboard回调，目前实现逻辑抽象到torch4keras中了，bert4torch使用方式和之前一样，参考https://github.com/Tongjilibo/bert4torch/blob/master/examples/tutorials/tutorials_small_tips.py

跑出来了，是acc和loss，目前支持看到f1或其他的吗？如果改，从哪里改？

**Comment by XuJianzhi** [comment]
波哥好。
出来的board显示的仅是train的指标，能否显示dev和test的指标？
是从这里加吗：https://github.com/Tongjilibo/torch4keras/blob/master/torch4keras/model.py#L238
加上类似：
perf_dev = metric_mapping(metric, func, output_dev, dev_y_)

**Comment by Tongjilibo** [comment]
可以显示，应该是不需要改包里面的代码的，同上所说，你在前面的评估后把对应的指标写到logs里面去，如下所示
```python
    def on_epoch_end(self, global_step, epoch, logs=None):
        val_acc = self.evaluate(valid_dataloader)
        test_acc = self.evaluate(test_dataloader)
        logs['val/acc'] = val_acc
        logs['test/acc'] = test_acc
```

**Comment by XuJianzhi** [comment]
我明白，但这样的结果是在日志中打印出来，，，能否让tensorboard上出现dev、test的情况？
（我通过修改model.compile的metrics已经让tensorboard上有我想要的指标在train集上的图。）

**Comment by Tongjilibo** [comment]
我可能没表达清楚，我的意思是上述这么操作应该就可以在tensorboard上记录了啊，要在tensorboard中显示dev和test的指标，就在前面对logs进行赋值，自带的tensorboard是默认记录所有logs里面的东西的，你运行[tuturial](https://github.com/Tongjilibo/bert4torch/blob/master/examples/tutorials/tutorials_small_tips.py)应该就能看出来有dev和test的指标吧，就像下面这样调用（我昨晚跑了下是可以在tensorboard中显示dev和test的啊），如果不能显示就是我实现的有问题，否则就无法满足用户自定义的需求了（bert4torch就不够通用了）
```python
class Evaluator(Callback):
    """评估与保存
    """
    def __init__(self):
        self.best_val_acc = 0.

    def on_epoch_end(self, global_step, epoch, logs=None):
        val_acc = self.evaluate(valid_dataloader)
        test_acc = self.evaluate(test_dataloader)
        logs['val/acc'] = val_acc  # 在此记录好，在Tensorboard中即可显示，注意Tensorboard需要在evaluator后声明
        logs['test/acc'] = test_acc
        if val_acc > self.best_val_acc:
            self.best_val_acc = val_acc
            # model.save_weights('best_model.pt')
        print(f'val_acc: {val_acc:.5f}, test_acc: {test_acc:.5f}, best_val_acc: {self.best_val_acc:.5f}\n')

        model.save_weights('last_model.pt', prefix=None)  # 保存模型权重
        model.save_steps_params('last_steps.pt')  # 保存训练进度参数，当前的epoch和step，断点续训使用
        torch.save(optimizer.state_dict(), 'last_optimizer.pt')  # 保存优化器，断点续训使用

if __name__ == '__main__':
    evaluator = Evaluator()
    model.fit(train_dataloader, epochs=10, steps_per_epoch=100, callbacks=[evaluator, Logger('test.log'), Tensorboard('./')])

```
你这么操作也是可以，但是这样就需要对框架进行修改

**Comment by XuJianzhi** [comment]
明白了，，之前看错了，，感谢！
